activerecord/lib/active_record/associations/association.rb

```diff
@@ -228,15 +229,72 @@ def find_target
             target_scope.merge!(as.scope(self))
           end
 
-          binds = AssociationScope.get_bind_values(owner, reflection.chain)
-          sc.execute(binds, klass.connection) do |record|
-            set_inverse_instance(record)
-            if owner.strict_loading_n_plus_one_only? && reflection.macro == :has_many
-              record.strict_loading!
-            else
-              record.strict_loading!(false, mode: owner.strict_loading_mode)
+          if preloaded_records?
+            records = attempt_preload()
+          else
+            binds = AssociationScope.get_bind_values(owner, reflection.chain)
+            records = sc.execute(binds, klass.connection) do |record|
+              set_inverse_instance(record)
+              if owner.strict_loading_n_plus_one_only? && reflection.macro == :has_many
+                record.strict_loading!
+              else
+                record.strict_loading!(false, mode: owner.strict_loading_mode)
+              end
+            end
+
+            # Setup the siblings for the preload tooling if it
+            # gets turned on in the future, or is needed by some other.
+            # tooling.
+            records.each do |record|
+              record._create_load_tree(siblings: records, parent: owner, association_name: reflection.name)
             end
+
+            ActiveRecord::LoadTree.add_loaded_records(records)
           end
+          records
+        end
+
+        def preloaded_records?
+          return false unless owner.respond_to?(reflection.name) # Sanity check
+          return false if owner.strict_loading_n_plus_one_only? # If strict loading is on do not preload
+          return false unless ActiveRecord.dynamic_includes_enabled? # Auto includes off, no preloading
+          return false if owner._load_tree.siblings.length <= 1 # no siblings so lets not go through the preload tooling,
+          # single records can sometime have optimizations not available
+          # to preloaders
+          true
+        end
+
+        def attempt_preload
+          available_records = [owner._load_tree.siblings]
+          available_records = Array.wrap(target) + available_records
+          available_records = available_records.flatten.uniq
```

Comment:
I am not sure if these available records are actually being used, or rather when. It seems that there are a few guards around how to use this part of the api.